File: src/cruart.jl
module cruart
using Gumbo, CSV, FileWatching, DataFrames, Lazy, TimeZones, Dates

const data_pfx = "data"
const data_fdr = data_pfx * "/"

function run()
    unwatch_folder(data_pfx)
    println("Waiting for data...")
    while true
        new_file = watch_folder(data_pfx)[1]
        if isfile(data_fdr * new_file)
            main(data_fdr * new_file)
        end
    end
end

function parse_cruart_date(str)
    r = r"/"
    m = match(r,str)
    month = parse(Int,str[1:m.offset-1])
    m2 = match(r,str,m.offset+1)
    day = parse(Int,str[m.offset+1:m2.offset-1])
    year = parse(Int,str[length(str)-3:length(str)])
    (year, month, day)
end

function main(filename)
    tz = TimeZone("America/Detroit")
    utc = TimeZone("UTC")
    html_data = try
        parsehtml(read(filename, String))
    catch ex
        println(ex)
        nothing
    end
    if html_data == nothing
        return
    end
    pages = length(html_data.root.children[2].children)

    results_df = DataFrame(trainee = String[], minutes = Int[], training_date = String[], training_pos = String[])

    for page in 1:pages
        for row in html_data.root.children[2].children[page].children[1].children
            if is_training_row(row)
                trainee = row.children[4].children[1].children[1].text
                trainee = trainee[length(trainee)-1:length(trainee)]
                minutes = parse_minutes(row.children[10].children[1].children[1].text)
                date = row.children[7].children[1].children[1].text
                time = row.children[8].children[1].children[1].text
                dt = ZonedDateTime(parse_cruart_date(date)...,
                                   parse(Int,time[1:2]),
                                   parse(Int,time[4:5]),
                                   utc)
                local_dt = Dates.format(astimezone(dt,tz), "mm/dd/yyyy")
                pos = row.children[5].children[1].children[1].text
                if pos == "RCIC" || pos == "CICA"
                    continue
                else
                    push!(results_df, (trainee, minutes, local_dt, pos))
                end
            end
        end
    end

    group_df = @> begin
        results_df
        DataFrames.groupby([:trainee, :training_date, :training_pos])
        aggregate(sum)
        rename((:minutes_sum => :minutes))
    end

    outfile = filename[match(r"\/",filename).offset+1:match(r"\.",filename).offset-1]
    try
        CSV.write(outfile * ".csv",group_df)
        println("Successfully wrote " * outfile * ".csv")
    catch ex
        println("Failed to write " * outfile * ".csv")
        println(ex)
        nothing
    end
end

function parse_minutes(hours_text)
    sep = match(r"\+",hours_text).offset
    hours = parse(Int,hours_text[1:sep-1])
    minutes = parse(Int,hours_text[sep+2:length(hours_text)])
    minutes + hours*60
end

function is_training_row(row)
    length(row.children) == 10 &&
        typeof(row.children[10].children[1].children[1]) == HTMLText &&
        row.children[6].children[1].children[1].text == "T"
end

run()
end # module

File: src/cruart.py
from bs4 import BeautifulSoup
from zoneinfo import ZoneInfo
from datetime import datetime, timezone
import pandas as pd
import re, sys

def is_training_row(row):
    return len(row) == 10 and \
        list(list(list(row.children)[5].children)[0].children)[0] == 'T'

def parse_minutes(text):
    sep = text.index("+")
    hours = int(text[0:sep])
    minutes = int(text[sep+1:])
    return (minutes + hours*60)

def parse_cruart_date(text):
    p = re.compile("(.*)/(.*)/(.*)")
    m = p.match(text)
    month = int(m.group(1))
    day = int(m.group(2))
    year = int(m.group(3))
    return (year,month,day)
    

def main(filename):
    columns=["trainee", "minutes", "training_date", "training_pos", "ojti", "t_date"]
    results = pd.DataFrame(columns = columns)
    
    tz = ZoneInfo("America/Detroit")
    with open(filename) as f:
        html_txt = f.read()
    html_data = BeautifulSoup(html_txt, "html.parser")

    tables = html_data.find_all("table")

    for table in tables:
        for row in table.find_all("tr"):
            if is_training_row(row):
                cols = row.find_all("td")
                trainee = cols[3].font.string[-2:]
                minutes = parse_minutes(cols[9].font.string)
                date = parse_cruart_date(cols[6].font.string)
                time = cols[7].font.string
                dt = datetime(*date, int(time[0:2]), int(time[3:5]), tzinfo=timezone.utc)
                local_dt = dt.astimezone(tz).strftime("%m/%d/%Y %H:%M")
                pos = cols[4].font.string
                ojti = cols[2].font.string[-2:]
                if pos == "RCIC" or pos == "CICA":
                    continue
                t_date = dt.strftime("%Y%m%d")
                df = pd.DataFrame([[trainee,minutes,local_dt,pos,ojti,t_date]], columns=columns)
                results = pd.concat([results,df],ignore_index=True)
    outfile = re.compile("(.*)\.htm").match(filename).group(1) + ".csv"
    results.to_csv(outfile, index=False)

main(sys.argv[1])
